packages/block-editor/src/components/block-tools/block-toolbar-popover.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import BlockToolbar from '../block-toolbar';
+import BlockPopover from '../block-popover';
+import useBlockToolbarPopoverProps from './use-block-toolbar-popover-props';
+import useSelectedBlockToolProps from './use-selected-block-tool-props';
+
+export default function BlockToolbarPopover( {
+	clientId,
+	__unstableContentRef,
+} ) {
+	const isTyping = useSelect(
+		( select ) => select( blockEditorStore ).isTyping(),
+		[]
+	);
+
+	const { capturingClientId, isInsertionPointVisible, lastClientId } =
+		useSelectedBlockToolProps( clientId );
+
+	const popoverProps = useBlockToolbarPopoverProps( {
+		contentElement: __unstableContentRef?.current,
+		clientId,
+	} );
+
+	return (
+		<BlockPopover
+			clientId={ capturingClientId || clientId }
+			bottomClientId={ lastClientId }
+			className={ classnames( 'block-editor-block-list__block-popover', {
+				'is-insertion-point-visible': isInsertionPointVisible,
+				'is-hidden': isTyping, // Leave the toolbar in the DOM so it can be focused at the same roving tabindex it was previously at
```

Comment:
Wow, this is a big change. I don't understand why this is needed now and not previously? This will impact performance. Cc @youknowriad

Reply:
Agree with @ellatrix here, the type metric went from 34ms to 39ms in the last run and "typing" is the critical metric in the editor. So I'd rather we remove the toolbar entirely as well 👍

Reply:
Let's avoid changing behavior in this refactoring PR.

Reply:
Right, it's not only about BlockToolbar here, but Popover as well, which might be a "heavy" component for performance.

Reply:
I wasn't aware of how heavy that component was. I assumed showing/hiding it would be less system work than destroying/remounting it. 

The reason it was helpful to make this change is that we were trying to pass as few props down to block toolbar. If we revert back to the mount/unmount method, we have to track the tabindex before unmount, and restore that tabindex on remount and pass it to `<BlockToolbar />` since it's used by `<NavigableToolbar />`. This would make `<BlockToolbar />` look like:
```
function BlockToolbar( {
  hideDragHandle,
  variant,
  __experimentalInitialIndex,
  __experimentalOnIndexChange,
  focusOnMount,
})
```

By hiding instead of unmounting, the tabindex in the `BlockToolbar` doesn't change, so we don't have to do any work. I'm quite surprised that that this show/hide is a typing performance hit. It seems like showing/hiding with CSS should be much less intensive than building and remounting, and when typing, the block popover shouldn't need to be aware of anything. 

I'd like to look into why there's a performance hit here, as I want avoid passing those additional props, especially when the show/hide is accomplishing the same end goal with less code and a simpler mental model.

Reply:
A component that is mounted subscribes to stores, rerenders as the attributes changes... There might be some small gains to be made for sure but I don't know that we would be able to have the same gains that "not rendering" the components allows us to have while typing. 

It's not the operation of "hiding/showing" that is slow. It's the fact that the component stays there with its state...

Regardless of the reasons you wanted to keep in the DOM, it seems like a separate issue than what the current PR is trying to do (Refactor and restore the previous way of rendering block toolbars), so for me we should just restore the previous behavior and address that issue in its own dedicated PR.

Reply:
Is it OK to add these `__experimentalInitialIndex`, `__experimentalOnIndexChange`, `focusOnMount` props to `BlockToolbar` then? If so, I can make that change.

Reply:
If we can keep them as private props instead yeah. Prefixed things (props, functions...) shouldn't be used anymore.